wasm.go

```diff
@@ -1,66 +1,171 @@
+	ctx := config.Context
+	if ctx == nil {
+		ctx = context.Background()
 	}
-	var name string
-	if m.NameSection != nil {
-		name = m.NameSection.ModuleName
+	engine := config.Engine
+	if engine == nil {
+		engine = NewEngineInterpreter()
 	}
-	return &Module{wasm: m, name: name}, nil
+	return internalwasm.NewStore(ctx, engine.e)
 }
 
-// EncodeModule encodes the given module into a byte slice depending on the format of the implementation.
-// See binary.EncodeModule
-type EncodeModule func(m *Module) []byte
-
-var EncodeModuleBinary EncodeModule = func(m *Module) []byte {
-	return encodeModule(binary.EncodeModule, m)
+// ModuleConfig defines the WebAssembly 1.0 (MVP) module to instantiate.
+type ModuleConfig struct {
+	// Name defaults to what's decoded from the custom name section and can be overridden WithName.
+	// See https://www.w3.org/TR/wasm-core-1/#name-section%E2%91%A0
+	Name string
+	// Source is the WebAssembly 1.0 (MVP) text or binary encoding of the module.
+	Source []byte
 }
 
-func encodeModule(encoder internalwasm.EncodeModule, m *Module) []byte {
-	return encoder(m.wasm)
+// InstantiateModule instantiates the module namespace or errs if the configuration was invalid.
+//
+// Ex.
+//	exports, _ := wazero.InstantiateModule(wazero.NewStore(), &wazero.ModuleConfig{Source: wasm})
+//
+// Note: StoreConfig.Context is used for any WebAssembly 1.0 (MVP) Start Function.
+func InstantiateModule(store wasm.Store, module *ModuleConfig) (wasm.ModuleExports, error) {
+	internal, ok := store.(*internalwasm.Store)
+	if !ok {
+		return nil, fmt.Errorf("unsupported Store implementation: %s", store)
+	}
+	m, name, err := decodeModule(module)
+	if err != nil {
+		return nil, err
+	}
+	return internal.Instantiate(m, name)
 }
 
-// Module is a WebAssembly 1.0 (MVP) module decoded (DecodeModule) from a valid source.
-type Module struct {
-	name string
-	wasm *internalwasm.Module
+func decodeModule(module *ModuleConfig) (m *internalwasm.Module, name string, err error) {
+	if module.Source == nil {
+		err = errors.New("source == nil")
+		return
+	}
+
+	if len(module.Source) < 8 { // Ex. less than magic+version in binary or '(module)' in text
+		err = errors.New("invalid source")
+		return
+	}
+
+	// Peek to see if this is a binary or text format
+	if bytes.Equal(module.Source[0:4], binary.Magic) {
+		m, err = binary.DecodeModule(module.Source)
+	} else {
+		m, err = text.DecodeModule(module.Source)
+	}
+	if err != nil {
+		return
+	}
+
+	name = module.Name
+	if name == "" && m.NameSection != nil {
+		name = m.NameSection.ModuleName
+	}
+	return
 }
 
-// Name defaults to what's decoded from the custom name section and can be overridden WithName.
-// See https://www.w3.org/TR/wasm-core-1/#name-section%E2%91%A0
-func (m *Module) Name() string {
-	return m.name
+// HostModuleConfig are WebAssembly 1.0 (MVP) exports from the host bound to a module name used by InstantiateHostModule.
+type HostModuleConfig struct {
```

Comment:
This type will later have globals, memory and tables added to it. These are able to be added later without breaking API, hence this change being raised first.